google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -154,32 +163,91 @@ def reset
 
         def close
           @thread.kill if @thread
-          @all_sessions.each(&:delete_session)
-          @all_sessions = []
-          @session_queue = []
-          @transaction_queue = []
+          # Unblock all waiting threads
+          @closed = true
+          @resource.broadcast
+          # Delete all sessions
+          @mutex.synchronize do
+            @all_sessions.map do |s|
+              Thread.new do
```

Comment:
if there are 1000 sessions, this will create a 1000 threads. can you instead use some sort of a threadpool?
https://github.com/ruby-concurrency/concurrent-ruby/blob/master/doc/thread_pools.md